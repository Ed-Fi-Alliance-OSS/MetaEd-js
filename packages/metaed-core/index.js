// @flow

// Test helper -- *** where should this go? ***
export { MetaEdTextBuilder } from './test/MetaEdTextBuilder';


// Functions

export { deepFreeze, deepFreezeAssign, prependIndefiniteArticle } from './src/Utility';
export { startingFromFileLoadP } from './src/task/Pipeline';
export { createMetaEdFile } from './src/task/MetaEdFile';
export { loadCoreBufferedFiles, loadExtensionBufferedFiles } from './src/task/BufferFileLoader';
export { loadFiles } from './src/task/FileSystemFilenameLoader';
export { newState } from './src/State';
export { addEntity, getAllEntities, getAllTopLevelEntities, getEntity, getTopLevelCoreEntity, getEntitiesOfType, getAllEntitiesNoSimpleTypes } from './src/model/EntityRepository';
export { addProperty, getAllProperties, getPropertiesOfType, newPropertyIndex } from './src/model/property/PropertyRepository';
export { isReferenceProperty, isSharedProperty, allPropertyTypes } from './src/model/property/PropertyType';
export { allEntityModelTypes, allTopLevelEntityModelTypes, topLevelCoreEntityModelTypes } from './src/model/ModelType';

// Builders

export { AssociationBuilder } from './src/builder/AssociationBuilder';
export { AssociationExtensionBuilder } from './src/builder/AssociationExtensionBuilder';
export { AssociationSubclassBuilder } from './src/builder/AssociationSubclassBuilder';
export { ChoiceBuilder } from './src/builder/ChoiceBuilder';
export { CommonBuilder } from './src/builder/CommonBuilder';
export { CommonExtensionBuilder } from './src/builder/CommonExtensionBuilder';
export { DecimalTypeBuilder } from './src/builder/DecimalTypeBuilder';
export { DescriptorBuilder } from './src/builder/DescriptorBuilder';
export { DomainBuilder } from './src/builder/DomainBuilder';
export { DomainEntityBuilder } from './src/builder/DomainEntityBuilder';
export { DomainEntityExtensionBuilder } from './src/builder/DomainEntityExtensionBuilder';
export { DomainEntitySubclassBuilder } from './src/builder/DomainEntitySubclassBuilder';
export { EnumerationBuilder } from './src/builder/EnumerationBuilder';
export { IntegerTypeBuilder } from './src/builder/IntegerTypeBuilder';
export { InterchangeBuilder } from './src/builder/InterchangeBuilder';
export { NamespaceInfoBuilder } from './src/builder/NamespaceInfoBuilder';
export { SharedDecimalBuilder } from './src/builder/SharedDecimalBuilder';
export { SharedIntegerBuilder } from './src/builder/SharedIntegerBuilder';
export { SharedStringBuilder } from './src/builder/SharedStringBuilder';
export { StringTypeBuilder } from './src/builder/StringTypeBuilder';

// Flow types

export type { State } from './src/State';
export type { InputDirectory } from './src/task/FileSystemFilenameLoader';
export type { FileSet, MetaEdFile } from './src/task/MetaEdFile';
export type { FileIndex } from './src/task/FileIndex';
export type { Validator } from './src/validator/Validator';
export type { ValidationFailure } from './src/validator/ValidationFailure';
export type { Enhancer } from './src/enhancer/Enhancer';
export type { EnhancerResult } from './src/enhancer/EnhancerResult';
export type { Generator } from './src/generator/Generator';
export type { GeneratorResult } from './src/generator/GeneratorResult';
export type { GeneratedOutput } from './src/generator/GeneratedOutput';
export type { MetaEdEnvironment, PluginEnvironment, SemVer } from './src/MetaEdEnvironment';
export type { MetaEdPlugin, PluginManifest } from './src/plugin/PluginTypes';

// Properties
export type { AssociationProperty } from './src/model/property/AssociationProperty';
export type { BooleanProperty } from './src/model/property/BooleanProperty';
export type { ChoiceProperty } from './src/model/property/ChoiceProperty';
export type { CommonProperty } from './src/model/property/CommonProperty';
export type { CurrencyProperty } from './src/model/property/CurrencyProperty';
export type { DateProperty } from './src/model/property/DateProperty';
export type { DecimalProperty } from './src/model/property/DecimalProperty';
export type { DescriptorProperty } from './src/model/property/DescriptorProperty';
export type { DomainEntityProperty } from './src/model/property/DomainEntityProperty';
export type { DurationProperty } from './src/model/property/DurationProperty';
export type { EntityProperty, HasReferencedEntity } from './src/model/property/EntityProperty';
export type { EnumerationProperty } from './src/model/property/EnumerationProperty';
export type { InlineCommonProperty } from './src/model/property/InlineCommonProperty';
export type { IntegerProperty } from './src/model/property/IntegerProperty';
export type { MergedProperty } from './src/model/property/MergedProperty';
export type { PercentProperty } from './src/model/property/PercentProperty';
export type { PropertyIndex } from './src/model/property/PropertyRepository';
export type { PropertyType } from './src/model/property/PropertyType';
export type { ReferentialProperty } from './src/model/property/ReferentialProperty';
export type { SchoolYearEnumerationProperty } from './src/model/property/SchoolYearEnumerationProperty';
export type { SharedDecimalProperty } from './src/model/property/SharedDecimalProperty';
export type { SharedIntegerProperty } from './src/model/property/SharedIntegerProperty';
export type { SharedShortProperty } from './src/model/property/SharedShortProperty';
export type { SharedStringProperty } from './src/model/property/SharedStringProperty';
export type { ShortProperty } from './src/model/property/ShortProperty';
export type { SimpleProperty } from './src/model/property/SimpleProperty';
export type { SharedSimple } from './src/model/SharedSimple';
export type { StringProperty } from './src/model/property/StringProperty';
export type { TimeProperty } from './src/model/property/TimeProperty';
export type { YearProperty } from './src/model/property/YearProperty';

// Entities
export type { AbstractEntity } from './src/model/AbstractEntity';
export type { Association } from './src/model/Association';
export type { AssociationExtension } from './src/model/AssociationExtension';
export type { AssociationSubclass } from './src/model/AssociationSubclass';
export type { Choice } from './src/model/Choice';
export type { Common } from './src/model/Common';
export type { CommonExtension } from './src/model/CommonExtension';
export type { DecimalType } from './src/model/DecimalType';
export type { Descriptor } from './src/model/Descriptor';
export type { Domain } from './src/model/Domain';
export type { DomainEntity } from './src/model/DomainEntity';
export type { DomainEntityExtension } from './src/model/DomainEntityExtension';
export type { DomainEntitySubclass } from './src/model/DomainEntitySubclass';
export type { DomainItem } from './src/model/DomainItem';
export type { EntityRepository } from './src/model/EntityRepository';
export type { Enumeration } from './src/model/Enumeration';
export type { EnumerationItem } from './src/model/EnumerationItem';
export type { IntegerType } from './src/model/IntegerType';
export type { Interchange } from './src/model/Interchange';
export type { InterchangeExtension } from './src/model/InterchangeExtension';
export type { InterchangeItem } from './src/model/InterchangeItem';
export type { MapTypeEnumeration } from './src/model/MapTypeEnumeration';
export type { ModelBase } from './src/model/ModelBase';
export type { ModelType } from './src/model/ModelType';
export type { NamespaceInfo } from './src/model/NamespaceInfo';
export type { SchoolYearEnumeration } from './src/model/SchoolYearEnumeration';
export type { SharedDecimal } from './src/model/SharedDecimal';
export type { SharedInteger } from './src/model/SharedInteger';
export type { SharedString } from './src/model/SharedString';
export type { SourceMap } from './src/model/SourceMap';
export type { StringType } from './src/model/StringType';
export type { Subdomain } from './src/model/Subdomain';
export type { TopLevelEntity } from './src/model/TopLevelEntity';

// Null Objects
export { NoMapTypeEnumeration } from './src/model/MapTypeEnumeration';
export { NoTopLevelEntity } from './src/model/TopLevelEntity';

// Factories
export { newMetaEdEnvironment, newPluginEnvironment } from './src/MetaEdEnvironment';
export { newMetaEdPlugin } from './src/plugin/PluginTypes';

export { newAssociationProperty } from './src/model/property/AssociationProperty';
export { newBooleanProperty } from './src/model/property/BooleanProperty';
export { newChoiceProperty } from './src/model/property/ChoiceProperty';
export { newCommonProperty } from './src/model/property/CommonProperty';
export { newCurrencyProperty } from './src/model/property/CurrencyProperty';
export { newDateProperty } from './src/model/property/DateProperty';
export { newDecimalProperty } from './src/model/property/DecimalProperty';
export { newDescriptorProperty } from './src/model/property/DescriptorProperty';
export { newDomainEntityProperty } from './src/model/property/DomainEntityProperty';
export { newDurationProperty } from './src/model/property/DurationProperty';
export { newEnumerationProperty } from './src/model/property/EnumerationProperty';
export { newInlineCommonProperty } from './src/model/property/InlineCommonProperty';
export { newIntegerProperty } from './src/model/property/IntegerProperty';
export { newMergedProperty } from './src/model/property/MergedProperty';
export { newPercentProperty } from './src/model/property/PercentProperty';
export { newSchoolYearEnumerationProperty } from './src/model/property/SchoolYearEnumerationProperty';
export { newSharedDecimalProperty } from './src/model/property/SharedDecimalProperty';
export { newSharedIntegerProperty } from './src/model/property/SharedIntegerProperty';
export { newSharedShortProperty } from './src/model/property/SharedShortProperty';
export { newSharedStringProperty } from './src/model/property/SharedStringProperty';
export { newShortProperty } from './src/model/property/ShortProperty';
export { newSimpleProperty } from './src/model/property/SimpleProperty';
export { newStringProperty } from './src/model/property/StringProperty';
export { newTimeProperty } from './src/model/property/TimeProperty';
export { newYearProperty } from './src/model/property/YearProperty';

export { newAbstractEntity } from './src/model/AbstractEntity';
export { newAssociation } from './src/model/Association';
export { newAssociationExtension } from './src/model/AssociationExtension';
export { newAssociationSubclass } from './src/model/AssociationSubclass';
export { newChoice } from './src/model/Choice';
export { newCommon } from './src/model/Common';
export { newCommonExtension } from './src/model/CommonExtension';
export { newDecimalType } from './src/model/DecimalType';
export { newDescriptor } from './src/model/Descriptor';
export { newDomain } from './src/model/Domain';
export { newDomainEntity } from './src/model/DomainEntity';
export { newDomainEntityExtension } from './src/model/DomainEntityExtension';
export { newDomainEntitySubclass } from './src/model/DomainEntitySubclass';
export { newDomainItem } from './src/model/DomainItem';
export { newEnumeration } from './src/model/Enumeration';
export { newEnumerationItem } from './src/model/EnumerationItem';
export { newInlineCommon } from './src/model/Common';
export { newIntegerType } from './src/model/IntegerType';
export { newInterchange } from './src/model/Interchange';
export { newInterchangeExtension } from './src/model/InterchangeExtension';
export { newInterchangeItem } from './src/model/InterchangeItem';
export { newMapTypeEnumeration } from './src/model/MapTypeEnumeration';
export { newNamespaceInfo } from './src/model/NamespaceInfo';
export { newSchoolYearEnumeration } from './src/model/SchoolYearEnumeration';
export { newSharedDecimal } from './src/model/SharedDecimal';
export { newSharedInteger } from './src/model/SharedInteger';
export { newSharedString } from './src/model/SharedString';
export { newSourceMap } from './src/model/SourceMap';
export { newStringType } from './src/model/StringType';
export { newSubdomain } from './src/model/Subdomain';
export { asTopLevelEntity, newTopLevelEntity } from './src/model/TopLevelEntity';
