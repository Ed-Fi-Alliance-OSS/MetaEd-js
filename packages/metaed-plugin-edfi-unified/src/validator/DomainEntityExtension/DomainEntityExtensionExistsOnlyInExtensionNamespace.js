// @flow
import type { MetaEdEnvironment, ValidationFailure } from '../../../../../packages/metaed-core/index';

// eslint-disable-next-line no-unused-vars
export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.entity.domainEntityExtension.forEach(entity => {
    if (!entity.namespaceInfo.isExtension) {
      failures.push({
        validatorName: 'DomainEntityExtensionExistsOnlyInExtensionNamespace',
        category: 'error',
        message: `DomainEntity additions '${entity.metaEdName}' is not valid in core namespace '${entity.namespaceInfo.namespace}`,
        sourceMap: entity.sourceMap.type,
        fileMap: null,
      });
    }
  });

  return failures;
}

