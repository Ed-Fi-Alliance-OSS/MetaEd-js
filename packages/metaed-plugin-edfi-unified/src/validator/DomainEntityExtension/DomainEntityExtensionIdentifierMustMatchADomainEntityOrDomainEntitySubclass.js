// @flow
import type { MetaEdEnvironment } from '../../../../../packages/metaed-core/src/MetaEdEnvironment';
import type { ValidationFailure } from '../../../../../packages/metaed-core/src/validator/ValidationFailure';

// eslint-disable-next-line no-unused-vars
export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.entity.domainEntityExtension.forEach(entity => {
    if (!metaEd.entity.domainEntity.has(entity.metaEdName) && !metaEd.entity.domainEntitySubclass.has(entity.metaEdName)) {
      failures.push({
        validatorName: 'DomainEntityExtensionIdentifierMustMatchADomainEntityOrDomainEntitySubclass',
        category: 'error',
        message: `Domain Entity additions '${entity.metaEdName}' does not match any declared Domain Entity or Domain Entity Subclass.`,
        sourceMap: entity.sourceMap.type,
        fileMap: null,
      });
    }
  });

  return failures;
}

