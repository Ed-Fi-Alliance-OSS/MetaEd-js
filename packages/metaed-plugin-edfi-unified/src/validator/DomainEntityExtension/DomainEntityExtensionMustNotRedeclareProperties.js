// @flow
import type { DomainEntity, DomainEntitySubclass, MetaEdEnvironment, ValidationFailure } from '../../../../metaed-core/index';
import { failExtensionPropertyRedeclarations } from '../ValidatorShared/FailExtensionPropertyRedeclarations';

// eslint-disable-next-line no-unused-vars
export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.entity.domainEntityExtension.forEach(domainEntityExtension => {
    const extendedEntity : DomainEntity | DomainEntitySubclass | void =
      metaEd.entity.domainEntity.get(domainEntityExtension.metaEdName) || metaEd.entity.domainEntitySubclass.get(domainEntityExtension.metaEdName);
    if (extendedEntity) {
      failExtensionPropertyRedeclarations('DomainEntityExtensionMustNotRedeclareProperties', domainEntityExtension, extendedEntity, failures);
    }
  });
  return failures;
}
