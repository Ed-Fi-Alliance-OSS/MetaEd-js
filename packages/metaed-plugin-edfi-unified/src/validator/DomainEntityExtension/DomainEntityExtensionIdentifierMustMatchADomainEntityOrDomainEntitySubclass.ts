import { MetaEdEnvironment, ValidationFailure, Namespace } from 'metaed-core';
import { getEntityForNamespaces } from 'metaed-core';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.namespace.forEach((namespace: Namespace) => {
    namespace.entity.domainEntityExtension.forEach(entity => {
      if (
        getEntityForNamespaces(entity.metaEdName, namespace.dependencies, 'domainEntity', 'domainEntitySubclass') == null
      ) {
        failures.push({
          validatorName: 'DomainEntityExtensionIdentifierMustMatchADomainEntityOrDomainEntitySubclass',
          category: 'error',
          message: `Domain Entity additions '${
            entity.metaEdName
          }' does not match any declared Domain Entity or Domain Entity Subclass.`,
          sourceMap: entity.sourceMap.type,
          fileMap: null,
        });
      }
    });
  });

  return failures;
}
