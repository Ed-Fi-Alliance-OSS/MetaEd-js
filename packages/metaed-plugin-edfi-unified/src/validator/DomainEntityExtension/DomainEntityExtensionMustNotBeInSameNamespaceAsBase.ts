import { MetaEdEnvironment, ValidationFailure } from 'metaed-core';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.namespace.forEach(namespace => {
    namespace.entity.domainEntityExtension.forEach(entity => {
      if (entity.baseEntityNamespaceName === namespace.namespaceName) {
        failures.push({
          validatorName: 'DomainEntityExtensionMustNotBeInSameNamespaceAsBase',
          category: 'error',
          message: `Domain Entity additions '${entity.metaEdName}' cannot be in the same namespace as its base entity.`,
          sourceMap: entity.sourceMap.metaEdName,
          fileMap: null,
        });
      }
    });
  });
  return failures;
}
