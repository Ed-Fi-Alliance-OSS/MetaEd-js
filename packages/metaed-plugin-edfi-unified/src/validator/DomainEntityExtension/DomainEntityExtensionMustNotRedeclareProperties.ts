import { MetaEdEnvironment, ValidationFailure, TopLevelEntity } from 'metaed-core';
import { getEntityForNamespaces } from 'metaed-core';
import { failExtensionPropertyRedeclarations } from '../ValidatorShared/FailExtensionPropertyRedeclarations';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.namespace.forEach(namespace => {
    namespace.entity.domainEntityExtension.forEach(domainEntityExtension => {
      const extendedEntity: TopLevelEntity | null = getEntityForNamespaces(
        domainEntityExtension.metaEdName,
        namespace.dependencies,
        'domainEntity',
        'domainEntitySubclass',
      ) as TopLevelEntity | null;

      if (extendedEntity != null) {
        failExtensionPropertyRedeclarations(
          'DomainEntityExtensionMustNotRedeclareProperties',
          domainEntityExtension,
          extendedEntity,
          failures,
        );
      }
    });
  });
  return failures;
}
