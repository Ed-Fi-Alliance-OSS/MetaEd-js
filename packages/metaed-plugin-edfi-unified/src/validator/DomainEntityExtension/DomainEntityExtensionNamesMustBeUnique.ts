// SPDX-License-Identifier: Apache-2.0
// Licensed to the Ed-Fi Alliance under one or more agreements.
// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
// See the LICENSE and NOTICES files in the project root for more information.

import { MetaEdEnvironment, ValidationFailure, Namespace } from '@edfi/metaed-core';
import { getEntitiesOfTypeForNamespaces } from '@edfi/metaed-core';
import { generateValidationErrorsForDuplicates } from '../ValidatorShared/ErrorsForDuplicateMetaEdNames';

export function validate(metaEd: MetaEdEnvironment): ValidationFailure[] {
  const failures: ValidationFailure[] = [];

  metaEd.namespace.forEach((namespace: Namespace) => {
    failures.push(
      ...generateValidationErrorsForDuplicates(
        getEntitiesOfTypeForNamespaces([namespace], 'domainEntityExtension'),
        'DomainEntityExtensionNamesMustBeUnique',
      ),
    );
  });

  return failures;
}
