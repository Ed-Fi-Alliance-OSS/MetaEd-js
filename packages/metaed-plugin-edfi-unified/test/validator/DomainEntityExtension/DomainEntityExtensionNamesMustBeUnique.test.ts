// SPDX-License-Identifier: Apache-2.0
// Licensed to the Ed-Fi Alliance under one or more agreements.
// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
// See the LICENSE and NOTICES files in the project root for more information.

import { newMetaEdEnvironment, MetaEdTextBuilder, DomainEntityExtensionBuilder, NamespaceBuilder } from '@edfi/metaed-core';
import { MetaEdEnvironment, ValidationFailure } from '@edfi/metaed-core';
import { validate } from '../../../src/validator/DomainEntityExtension/DomainEntityExtensionNamesMustBeUnique';

describe('when entities in same namespace have different names', (): void => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  let failures: ValidationFailure[];
  let coreNamespace: any = null;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('EdFi')
      .withStartDomainEntityExtension('ValidName1')
      .withBooleanProperty('PropertyName', 'doc', true, false)
      .withEndDomainEntityExtension()

      .withStartDomainEntityExtension('ValidName2')
      .withBooleanProperty('PropertyName', 'doc', true, false)
      .withEndDomainEntityExtension()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new DomainEntityExtensionBuilder(metaEd, []));

    coreNamespace = metaEd.namespace.get('EdFi');
    failures = validate(metaEd);
  });

  it('should build two domainEntityExtension', (): void => {
    expect(coreNamespace.entity.domainEntityExtension.size).toBe(2);
  });

  it('should have no validation failures', (): void => {
    expect(failures).toHaveLength(0);
  });
});

describe('when entities in same namespace have identical names', (): void => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const entityName = 'EntityName';
  let failures: ValidationFailure[];
  let coreNamespace: any = null;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('EdFi')
      .withStartDomainEntityExtension(entityName)
      .withBooleanProperty('PropertyName', 'doc', true, false)
      .withEndDomainEntityExtension()

      .withStartDomainEntityExtension(entityName)
      .withBooleanProperty('PropertyName', 'doc', true, false)
      .withEndDomainEntityExtension()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new DomainEntityExtensionBuilder(metaEd, []));

    coreNamespace = metaEd.namespace.get('EdFi');
    failures = validate(metaEd);
  });

  it('should build one domainEntityExtension because TopLevelEntityBuilder will not let it get that far', (): void => {
    expect(coreNamespace.entity.domainEntityExtension.size).toBe(1);
  });

  it('should have no validation failures because of TopLevelEntityBuilder', (): void => {
    expect(failures).toHaveLength(0);
  });
});

describe('when domainEntityExtensions in separate dependency-linked namespaces have identical names', (): void => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const entityName = 'EntityName';
  let failures: ValidationFailure[];
  let coreNamespace: any = null;
  let extensionNamespace: any = null;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('EdFi')
      .withStartDomainEntityExtension(entityName)
      .withBooleanProperty('PropertyName', 'doc', true, false)
      .withEndDomainEntityExtension()
      .withEndNamespace()

      .withBeginNamespace('Extension', 'Extension')
      .withStartDomainEntityExtension(entityName)
      .withBooleanProperty('PropertyName', 'doc', true, false)
      .withEndDomainEntityExtension()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new DomainEntityExtensionBuilder(metaEd, []));

    coreNamespace = metaEd.namespace.get('EdFi');
    extensionNamespace = metaEd.namespace.get('Extension');
    extensionNamespace.dependencies.push(coreNamespace);

    failures = validate(metaEd);
  });

  it('should build one domainEntityExtension in core namespace', (): void => {
    expect(coreNamespace.entity.domainEntityExtension.size).toBe(1);
  });

  it('should build one domainEntityExtension in extension namespace', (): void => {
    expect(extensionNamespace.entity.domainEntityExtension.size).toBe(1);
  });

  it('should have no validation failures', (): void => {
    expect(failures).toHaveLength(0);
  });
});

describe('when domainEntityExtensions in non-dependency-linked namespaces have identical names', (): void => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const entityName = 'EntityName';
  let failures: ValidationFailure[];
  let coreNamespace: any = null;
  let extensionNamespacea: any = null;
  let extensionNamespaceb: any = null;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('EdFi')
      .withStartDomainEntityExtension('NotADuplicate')
      .withBooleanProperty('PropertyName', 'doc', true, false)
      .withEndDomainEntityExtension()
      .withEndNamespace()

      .withBeginNamespace('Extensiona', 'Extensiona')
      .withStartDomainEntityExtension(entityName)
      .withBooleanProperty('PropertyName', 'doc', true, false)
      .withEndDomainEntityExtension()
      .withEndNamespace()

      .withBeginNamespace('Extensionb', 'Extensionb')
      .withStartDomainEntityExtension(entityName)
      .withBooleanProperty('PropertyName', 'doc', true, false)
      .withEndDomainEntityExtension()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new DomainEntityExtensionBuilder(metaEd, []));

    coreNamespace = metaEd.namespace.get('EdFi');
    extensionNamespacea = metaEd.namespace.get('Extensiona');
    extensionNamespaceb = metaEd.namespace.get('Extensionb');
    extensionNamespacea.dependencies.push(coreNamespace);
    extensionNamespaceb.dependencies.push(coreNamespace);

    failures = validate(metaEd);
  });

  it('should build one core domainEntityExtension', (): void => {
    expect(coreNamespace.entity.domainEntityExtension.size).toBe(1);
  });

  it('should build one extension1 domainEntityExtension', (): void => {
    expect(extensionNamespacea.entity.domainEntityExtension.size).toBe(1);
  });

  it('should build one extension2 domainEntityExtension', (): void => {
    expect(extensionNamespaceb.entity.domainEntityExtension.size).toBe(1);
  });

  it('should have no validation failures', (): void => {
    expect(failures).toHaveLength(0);
  });
});
