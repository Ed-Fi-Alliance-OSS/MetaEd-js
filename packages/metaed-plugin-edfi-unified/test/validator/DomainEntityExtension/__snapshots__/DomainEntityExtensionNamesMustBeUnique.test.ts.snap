// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when domainEntityExtensions in separate dependency-linked namespaces have identical names should have validation failures for each entity 1`] = `"Domain Entity Extension named EntityName is a duplicate declaration of that name."`;

exports[`when domainEntityExtensions in separate dependency-linked namespaces have identical names should have validation failures for each entity 2`] = `
Object {
  "column": 2,
  "line": 9,
  "tokenText": "Domain Entity",
}
`;

exports[`when domainEntityExtensions in separate dependency-linked namespaces have identical names should have validation failures for each entity 3`] = `"Domain Entity Extension named EntityName is a duplicate declaration of that name."`;

exports[`when domainEntityExtensions in separate dependency-linked namespaces have identical names should have validation failures for each entity 4`] = `
Object {
  "column": 2,
  "line": 2,
  "tokenText": "Domain Entity",
}
`;
